### PR TITLE
chore(pre-commit): Add pyupgrade and pycln hooks

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

I thought there was merit in "keeping up with the Joneses"  in terms of ensuring our Python code is inline with the evolution of the language.

[pyupgrade](https://github.com/asottile/pyupgrade) is a tool (and pre-commit hook) to automatically upgrade syntax for newer versions of the language. This PR adds said pre-commit hook (in addition to `pycln` which removes unused imports) to help ensure your code style format is fresh. 

Apologies for the scale of the change. I wasn't able to find any setting in `pyupgrade` to make this more digestible, unless we opted to incrementally set the minimum Python requirement.

### TESTING INSTRUCTIONS

CI.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -15,14 +15,29 @@
...
+        args:
+          - --keep-runtime-typing
+          - --py38-plus
+  - repo: https://github.com/hadialqattan/pycln
+    rev: v2.1.2
+    hooks:
+      - id: pycln
+        args:
+          - --disable-all-dunder-policy
+          - --exclude=superset/config.py
```
Comment (john-bodley) [inline .pre-commit-config.yaml:35]: `pycln` chokes on `module not found` errors when we try importing the default configs.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -15,14 +15,26 @@
...
+    hooks:
+      - id: auto-walrus
+  - repo: https://github.com/asottile/pyupgrade
+    rev: v3.4.0
+    hooks:
+      - id: pyupgrade
+        args:
+          - --keep-runtime-typing
+          - --py38-plus
+  - repo: https://github.com/hadialqattan/pycln
```
[the remark this answers is not in the document]
Comment (john-bodley) [reply .pre-commit-config.yaml:28]: @michael-s-molina yes. Per the `--py39-plus` CLI argument it ensures that the changes are compatible with Python versions >= 3.9 which is the minimum version we currently support.

**Review by michael-s-molina — APPROVED** [review]
LGTM. Thank you very much for this improvement @john-bodley! It will contribute a lot to code standardization.

Reviewing this PR made me aware of how much untyped structures (`dict`) we have in the codebase and the consequences of that in terms of bugs and duplicated logic 😟 

I'm also going to ping @villebro @eschutho @jfrag1 in case they have extensions that may be affected by this PR.

File: superset/connectors/sqla/models.py [hunk 3]
```diff
@@ -869,8 +858,8 @@ def get_sqla_table(self) -> TableClause:
         return tbl
 
     def get_from_clause(
-        self, template_processor: Optional[BaseTemplateProcessor] = None
-    ) -> Tuple[Union[TableClause, Alias], Optional[str]]:
+        self, template_processor: BaseTemplateProcessor | None = None
+    ) -> tuple[TableClause | Alias, str | None]:
```
[the remark this answers is not in the document]
Comment (john-bodley) [reply superset/connectors/sqla/models.py:862]: Yup. The term `Optional` is such a misnomer and a bad choice to represent something that could be `None`. The [PEP-604 typing rewrites](https://github.com/asottile/pyupgrade#pep-604-typing-rewrites) currently only apply to those files which contain the following, 

```
from __future__ import annotations
```

though will become ubiquitous after we drop support for Python 3.9.

**Pull request merged**